feat: mypy for all type check

yihong0618 opened the pull request:
# Summary

This maybe a huge Pull Request, the main goal is to make all the `mypy` type check passed to fix the potential bugs
So its still WIP and will be draft. and try to close https://github.com/langgenius/dify/issues/10928

the first steps had beed done. -> let `mypy` work in `api` folder.
And we fixed some of them,

you can check by

- gh pr checkout 10921
- when found  small bug fix in main branch first
- install mypy
- cd api && mypy .

- [x] make it work for mypy
- [x] add ci for it
- [x] always merge main for the latest code
- [x] fixed them folder by folder, maybe type by type

for now after this commit it still has 4261 errors

some details update here:

- https://github.com/pallets-eco/flask-sqlalchemy/issues/1327

![image](https://github.com/user-attachments/assets/88ed3cd1-15a2-41f0-8a0d-a4b540b9ec91)

@laipz8200 @crazywoola 


> [!Tip]
> Close issue syntax: `Fixes #<issue number>` or `Resolves #<issue number>`, see [documentation](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) for more details.


# Screenshots

<table>
  <tr>
  <td>Before: </td>
  <td>After: </td>
  </tr>
  <tr>
  <td>...</td>
  <td>...</td>
  </tr>
</table>

# Checklist

> [!IMPORTANT]  
> Please review the checklist below before submitting your pull request.

- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [x] I understand that this PR may be closed in case there was no previous discussion or issues. (This doesn't apply to typos!)
- [x] I've added a test for each change that was introduced, and I tried as much as possible to make a single atomic change.
- [x] I've updated the documentation accordingly.
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

yihong0618 committed:
feat: mypy for all type check

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

laipz8200 commented:
It looks like a great piece of work! Just so you know, due to some historical reasons, the type annotations in the current codebase are not always accurate. Also, extra caution is needed when dealing with dict types from requests or the database.

If you need any assistance, feel free to reach out to me.

laipz8200 commented:
Additionally, according to our [Contribution Guide](https://github.com/langgenius/dify/blob/main/CONTRIBUTING.md), I recommend linking an issue to this PR to clarify your objective and avoid others submitting duplicate PRs.

yihong0618 commented:
> It looks like a great piece of work! Just so you know, due to some historical reasons, the type annotations in the current codebase are not always accurate. Also, extra caution is needed when dealing with dict types from requests or the database.
> 
> If you need any assistance, feel free to reach out to me.

yes, maybe you can also push commit to this too, for the fixing them all, and I would create an issue for it

yihong0618 committed:
fix: merge main and 4195 errors left

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

laipz8200 commented:
For Flask-SQLAlchemy, I’d like to share our plan with you:

1. Replace the direct use of `db.session` with `Session` from `sqlalchemy.orm` for queries and updates.
2. In a future major release, we plan to have database models directly inherit from SQLAlchemy’s base class.
3. Finally, we aim to use SQLAlchemy directly instead of Flask-SQLAlchemy.

As a result, when encountering type-related issues, we can proceed with change 1 and temporarily ignore the errors caused by change 2.

yihong0618 commented:
> For Flask-SQLAlchemy, I’d like to share our plan with you:
> 
> 1. Replace the direct use of `db.session` with `Session` from `sqlalchemy.orm` for queries and updates.
> 2. In a future major release, we plan to have database models directly inherit from SQLAlchemy’s base class.
> 3. Finally, we aim to use SQLAlchemy directly instead of Flask-SQLAlchemy.
> 
> As a result, when encountering type-related issues, we can proceed with change 1 and temporarily ignore the errors caused by change 2.

yes, this time I also choose ignore it, thanks for the info

bowenliang123 commented:
With my -1 for this PR.

- mypy for static type check brings little benefits in project
- the in-line suppressions puzzles the developers, for what rule it's violating or how we could further fix it.
- no proper auto fixes tool to align with the mypy rules
- too many suppressions in this PR
- for the field types, suggest to use Pydantic which runs check in runtime

yihong0618 commented:
> With my -1 for this PR.
> 
> * mypy for static type check brings little benefits in project
> * the in-line suppressions puzzles the developers, for what rule it's violating or how we could further fix it.
> * no proper auto fixes tool to align with the mypy rules
> * too many suppressions in this PR
> * for the field types, suggest to use Pydantic which runs check in runtime

some are right, but for now, I checked many of the bugs caused by the type, if not fix it by now, the code base grows, it will never be fixed.

yihong0618 committed:
fix: tiny fix and make ci happy

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: make lockfile happy

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 commented:
> > With my -1 for this PR.
> > 
> > * mypy for static type check brings little benefits in project
> > * the in-line suppressions puzzles the developers, for what rule it's violating or how we could further fix it.
> > * no proper auto fixes tool to align with the mypy rules
> > * too many suppressions in this PR
> > * for the field types, suggest to use Pydantic which runs check in runtime
> 
> some are right, but for now, I checked many of the bugs caused by the type, if not fix it by now, the code base grows, it will never be fixed.

but this part I totally agree maybe even all the code happy, it should not in the CI, its too still some pain for everyone work on the mypy, so keep it on develop side maybe better 
> the in-line suppressions puzzles the developers, for what rule it's violating or how we could further fix it.
no proper auto fixes tool to align with the mypy rules

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
fix: wip

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: lint

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
fix: wip and less than 4000 errors 3908 now

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: all import-untyped done

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: wip 3608 left

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
fix: wip 3497 left

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
chore: poetry lock update

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: wip 3372 left

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
chore: lock file

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: wip and will change to python3.12

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
fix: drop conflict

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: lock version

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: wip

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
fix: merge main

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: wip

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
fix: wip

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
fix: Xinference api_key can be None

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: lint

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: wip 3145 left

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
fix: wip and ignores tools and models

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

laipz8200 committed:
chore(pyproject.toml): Move mypy configurations to `mypy.ini`

1. Move config from `pyproject.toml` to `mypy.ini`
2. Exclude `tests/` and `migrations/`

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
chore(*): Remove variable types

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix(graph_engine): Fix type error under `graph_engine`

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix(app_dsl_service): Check if tenant_id is None before create App

Signed-off-by: -LAN- <laipz8200@outlook.com>

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
fix: merge

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: update lock

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: update ruff check

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: still wip

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: wip

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
fix: wip

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: wip 1336 left

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: Dict -> dict

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: feishu and 1256 left

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: wip

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: wip 1180 left

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
fix: wip 1115 left

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
fix: vdb error

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
fix: wip

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
chore

yihong0618 committed:
fix: wip 1100+ left....

yihong0618 committed:
Merge branch 'hy/type_hints_all' of https://github.com/yihong0618/dify into hy/type_hints_all

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
fix: less than 1000 -> 999 continue to work hope can be done every day 100

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
fix: on progress 886 left

yihong0618 commented:
Less than 1000 still working on it

yihong0618 committed:
fix: wip 740 left

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: process 695 left

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
chore: merge main

yihong0618 committed:
fix: less than 500 now

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 commented:
less than 500 now

yihong0618 committed:
fix: ruff check

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: wip less than 400 now

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
fix: wip less than 400 errors 393 now

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: ruff lint

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
fix: merge

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: lock

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: wip 340 left

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
fix: less than 300 now

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 commented:
300 left

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
fix: less than 200 now

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 commented:
less than 200 now!

laipz8200 commented:
> less than 200 now!

Cool!

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
fix: wip

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: error

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix

yihong0618 committed:
fix: less that 100 now

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 commented:
98 now

laipz8200 commented:
wow

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
fix: merge main

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
Revert "fix: merge"

This reverts commit ab55b4ece6ded114c81d09d0df68ddfb14e96958.

yihong0618 committed:
fix: wip 65 left

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
fix: wip 30 left

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: ci import error

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: lint

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: done and some not sure type add FIXME

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 commented:
Done 
This is phase 1
for phase 2 
- drop `Any` as many as possible
- add type for the types empty var
- fix all the FIXME for type error

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

laipz8200 commented:
Awesome! I will review it in these two weeks.

yihong0618 commented:
> Awesome! I will review it in these two weeks.

seems need to keep updating...

laipz8200 commented:
We also need to add the CI configuration in this PR.

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
fix: it

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
ci: add mypy tests

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
ci: add test for mypy

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
fix: mypy version

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
Merge branch 'hy/type_hints_all' of https://github.com/yihong0618/dify into hy/type_hints_all

yihong0618 committed:
fix: lock

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
fix: must in api files

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
fix: yml...

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
fix: mypy

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

yihong0618 committed:
Merge branch 'main' into hy/type_hints_all

yihong0618 committed:
fix: super lint

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

laipz8200 commented on .github/workflows/api-tests.yml:
```diff
@@ -56,6 +56,12 @@ jobs:
+      - name: Run mypy
+        run: |
+          pushd api
+          poetry run -C api python -m mypy .
```
```suggestion
          poetry run python -m mypy .
```

yihong0618 committed:
ci: better mypy check

Signed-off-by: yihong0618 <zouzou0208@gmail.com>

laipz8200 committed:
fix: remove unnecessary type ignore comments from computed fields

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: update type hints for computed fields and headers

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: remove unnecessary type ignore comments and improve mypy command

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: update mypy command to use python module and remove unused python_executable setting

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: improve document syncing logic and handle missing documents

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: handle missing documents in retry indexing task and streamline error handling

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: enhance load method with overloads for better type hinting and simplify data loading in process_trace_tasks

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: refactor create_clusters function to use keyword arguments for clarity and maintainability

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: refactor update_clusters function to use named parameters for improved readability

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: remove unnecessary type ignore comment for CORS import

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: update parser_api_schema return type to Mapping for improved type safety

Sgned-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: replace computed_field decorators with property for improved clarity in PositionConfig

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: remove unnecessary type ignore comment for yaml import

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: update max_active_requests to use mapped_column for improved type safety

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: simplify return statement in AppDslService by removing unnecessary variable assignment

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: streamline count statement in ConversationService for improved readability

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: remove unnecessary type hint in MessageService for improved clarity

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: remove unnecessary return type hints in ProviderConfiguration and ModelProviderService for improved clarity

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
feat: add explore API resources for audio, completion, conversation, message, and workflow

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: enforce model_instance requirement in agent runners for improved reliability

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: enhance content handling in CotCompletionAgentRunner for improved message processing

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: simplify return statement in get_model_credentials for improved readability

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: remove unnecessary type casting in FunctionCallAgentRunner for improved clarity

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: improve message handling in CotAgentRunner by enforcing type checks and scratchpad validation

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: enforce content type validation in CotAgentRunner for improved error handling

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: enhance message content handling in AppGeneratorTTSPublisher for improved type validation

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: streamline conversation variable handling in AdvancedChatAppRunner for improved clarity

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: remove unnecessary type ignores in generate_response_converter for improved clarity

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: update type hints to use Mapping and Sequence for improved type safety

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: update inputs type hints to use Mapping for improved type safety

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: update type hints in BaseAppGenerator and WorkflowAppGenerator for improved type safety

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: improve invoke result handling in AppRunner for better type safety and clarity

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: update type hints in WorkflowBasedAppRunner and QueueNodeSucceededEvent for improved type safety

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: update type hints in AppGenerateEntity to use Sequence for files for improved type safety

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: update type hints in file_manager to use Mapping for improved type safety

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: replace @computed_field with @property for improved clarity and consistency in PositionConfig

Signed-off-by: -LAN- <laipz8200@outlook.com>

laipz8200 committed:
fix: replace @computed_field with @property in MultiModalPromptMessageContent for improved clarity

Signed-off-by: -LAN- <laipz8200@outlook.com>
